src/python/grpcio/tests/qps/benchmark_client.py

```diff
@@ -0,0 +1,195 @@
+
+"""Defines test client behaviors (UNARY/STREAMING) (SYNC/ASYNC)."""
+
+import abc
+import time
+try:
+  import Queue as queue  # Python 2.x
+except ImportError:
+  import queue  # Python 3
+
+import concurrent.futures
+from grpc.beta import implementations
+from src.proto.grpc.testing import messages_pb2
+from src.proto.grpc.testing import services_pb2
+from tests.unit import resources
+from tests.unit.beta import test_utilities
+
+_TIMEOUT = 60 * 60 * 24
+_SERVER_HOST_OVERRIDE = 'foo.test.google.fr'
+
+
+class BenchmarkClient:
+  """Benchmark client interface that exposes a non-blocking send_request()."""
+
+  __metaclass__ = abc.ABCMeta
+
+  def __init__(self, server, payload_config, use_ssl, hist):
+    # Create the stub
+    host, port = server.split(':')
+    port = int(port)
+    if use_ssl:
+      creds = implementations.ssl_channel_credentials(
+          resources.test_root_certificates())
+      channel = test_utilities.not_really_secure_channel(host, port, creds,
+                                                         _SERVER_HOST_OVERRIDE)
+    else:
+      channel = implementations.insecure_channel(host, port)
+
+    if payload_config.HasField('simple_params'):
+      self._generic = False
+      self._stub = services_pb2.beta_create_BenchmarkService_stub(channel)
+      payload = messages_pb2.Payload(
+          body='\0' * payload_config.simple_params.req_size)
+      self._request = messages_pb2.SimpleRequest(
+          payload=payload,
+          response_size=payload_config.simple_params.resp_size)
+    else:
+      self._generic = True
+      self._stub = implementations.generic_stub(channel)
+      self._request = '\0' * payload_config.bytebuf_params.req_size
+
+    self._hist = hist
+    self._response_callbacks = []
+
+  def add_response_callback(self, callback):
+    self._response_callbacks.append(callback)
+
+  @abc.abstractmethod
+  def send_request(self):
+    raise NotImplementedError()
+
+  def start(self):
+    pass
+
+  def stop(self):
+    pass
+
+  def _handle_response(self, query_time):
+    self._hist.add(query_time * 1e9)  # Report times in nanoseconds
+    for callback in self._response_callbacks:
+      callback(query_time)
+
+
+class UnarySyncBenchmarkClient(BenchmarkClient):
+
+  def __init__(self, server, payload_config, use_ssl, hist, max_rpcs):
+    super(UnarySyncBenchmarkClient, self).__init__(server, payload_config,
+                                                   use_ssl, hist)
+    self._pool = concurrent.futures.ThreadPoolExecutor(max_workers=max_rpcs)
+
+  def send_request(self):
+    # Send requests in seperate threads to support multiple outstanding rpcs
+    # (See src/proto/grpc/testing/control.proto)
+    self._pool.submit(self._dispatch_request)
+
+  def stop(self):
+    self._pool.shutdown(wait=True)
+    self._stub = None
+
+  def _dispatch_request(self):
+    start_time = time.time()
+    if self._generic:
+      self._stub.blocking_unary_unary('grpc.testing.BenchmarkService',
```

Comment:
Yes, but our performance testing framework doesn't support that, so maybe there's not much use in implementing that functionality.